File: Data Structures & Algorithms/permutation-string/submission-1.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        if (len(s1) > len(s2)):
            return False

        need = [0] * 26
        seen = [0] * 26
        for i in range(len(s1)):
            need[ord(s1[i]) - ord('a')] += 1
            seen[ord(s2[i]) - ord('a')] += 1
        
        if (need == seen):
            return True
        
        left = 0
        for right in range(len(s1), len(s2)):
            seen[ord(s2[right]) - ord('a')] += 1
            seen[ord(s2[left]) - ord('a')] -= 1
            left+= 1

            print(seen)
            if (need == seen):
                return True
        return False

File: Data Structures & Algorithms/is-palindrome/submission-1.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        int i = 0, j = s.size() - 1;
        char left, right;
        while (i < j) {
            left = s[i];
            right = s[j];

            if (!isalnum(left)) {
                i++;
                continue;
            }
            
            if (!isalnum(right)) {
                j--;
                continue;
            }

            if (tolower(s[i]) == tolower(s[j])) {
                i++;
                j--;
            }
            else {
                return false;
            }
        }
        return true;
    }
};
